### PR TITLE
AVX2 suppord

### PR DESCRIPTION
**Pull request by mkaluza**
I did another PR as I wanted to leave that branch and discussion intact

I cleaned it up (in my terms). I realize there are functions left in avx header. If you want them converted to macros, I can do it (although I haven't seen any speed difference, and macros in contrast to functions are completely undebugable, which is a problem sometimes).

If you want something else changed just tell me.

### CODE REVIEW

**Comment by JDarnley** [comment]
If I may make a comment.  I think you should change the "uninterleaved" in a comment added in 1cf9864 to "deinterleaved" because that is the term used elsewhere in the code.

File: src/dvbcsa_bs_avx2.h [hunk 1]
```diff
@@ -166,5 +150,59 @@ static inline void block_sbox_avx3(dvbcsa_bs_word_t *src, dvbcsa_bs_word_t *dst)
...
+/* block cipher 2-word load with byte-deinterleaving */
+// BS_LOAD_DEINTERLEAVE_8 replacement that only loads data that are already uninterleaved
+
+#define BS_LOAD_DEINTERLEAVE_8(ptr, var_lo, var_hi) \
+      {\
+      var_lo = _mm256_load_si256((ptr)); \
+      var_hi = _mm256_load_si256((ptr) + 1); \
+      }
+
+extern const uint16_t dvbcsa_block_sbox_perm[256];
```
Comment (glenvt18) [inline src/dvbcsa_bs_avx2.h]: use `static const uint16_t dvbcsa_block_sbox_perm[256]` (forward declaration)

File: src/dvbcsa_bs_block.c [hunk 2]
```diff
@@ -53,7 +53,7 @@
   LSB: sbox output
   MSB: permuted sbox output
 */
-static const uint16_t dvbcsa_block_sbox_perm[256] =
+const uint16_t dvbcsa_block_sbox_perm[256] =
```
Comment (glenvt18) [inline src/dvbcsa_bs_block.c]: this change is not needed

File: src/dvbcsa_bs_avx2.h [hunk 3]
```diff
@@ -166,5 +150,59 @@ static inline void block_sbox_avx3(dvbcsa_bs_word_t *src, dvbcsa_bs_word_t *dst)
...
+static inline void block_sbox_permute_interleave_avx(dvbcsa_bs_word_t *src, dvbcsa_bs_word_t *dst) {
```
Comment (glenvt18) [inline src/dvbcsa_bs_avx2.h]: I would name it `avx2_block_sbox_permute` or `block_sbox_permute_avx2`. It doesn't interleave anything.

File: src/dvbcsa_bs_avx2.h [hunk 4]
```diff
@@ -166,5 +150,59 @@ static inline void block_sbox_avx3(dvbcsa_bs_word_t *src, dvbcsa_bs_word_t *dst)
...
+		//unpack
+		a = _mm256_unpacklo_epi32(res1, res2);
+		b = _mm256_unpackhi_epi32(res1, res2);
+
+		a = _mm256_shuffle_epi8(res1, _mm256_set_epi8(15,13,11,9,7,5,3,1, 14,12,10,8,6,4,2,0, 15,13,11,9,7,5,3,1, 14,12,10,8,6,4,2,0));
+		b = _mm256_shuffle_epi8(res2, _mm256_set_epi8(15,13,11,9,7,5,3,1, 14,12,10,8,6,4,2,0, 15,13,11,9,7,5,3,1, 14,12,10,8,6,4,2,0));
+		res1 = _mm256_unpacklo_epi16(a, b);
+		res2 = _mm256_unpackhi_epi16(a, b);
+
+		_mm256_store_si256(dst + 2*j, res1);
```
Comment (glenvt18) [inline src/dvbcsa_bs_avx2.h]: style: `2*j -> 2 * j`

File: src/dvbcsa_bs_avx2.h [hunk 5]
```diff
@@ -166,5 +150,59 @@ static inline void block_sbox_avx3(dvbcsa_bs_word_t *src, dvbcsa_bs_word_t *dst)
 
 #define BLOCK_SBOX(in_buf, out_buf) block_sbox_avx2(in_buf, out_buf);
 
+/* block cipher 2-word load with byte-deinterleaving */
+// BS_LOAD_DEINTERLEAVE_8 replacement that only loads data that are already uninterleaved
```
Comment (glenvt18) [inline src/dvbcsa_bs_avx2.h]: I would change this to
`/* BS_LOAD_DEINTERLEAVE_8 stub: deinterleaving is done by BLOCK_SBOX_PERMUTE */`

**Comment by glenvt18** [comment]
@mkaluza Looks good.

Some notes.

*  Inline functions are OK.
*  Any experimental stuff (block_sbox_avxN) can't be merged into the master branch. Any alternative option must be supported by `configure.ac`. No commented blocks of code either. If you want to save this work for future use, keep it in your experimental branch. Master is not the appropriate place for that.

* Please stick to the libdvbcsa style: no tabs, 2 chars indents, no C++ comments, braces, etc.

**Comment by mkaluza** [comment]
@glenvt18 ok, I think that's all

**Comment by glenvt18** [comment]
@mkaluza Almost:) The patch is OK. I'd like to ask you to make some cosmetic changes to the commit message:

* `add` -> `Add`
* Replace tabs with spaces
* Add a short description of what has been done: AVX2 table lookup instruction, how it boosts performance compared to AVX2 + plain lookup, 16 bit fetch, deinterleaving done in the same loop, etc. Just shortly. It will help a viewer to catch the idea quickly.

**Comment by glenvt18** [comment]
@mkaluza Found a small issue:
```C
#ifndef DVBCSA_AVX_H_
#define DVBCSA_AVX_H_
```

should be `DVBCSA_AVX2_H_`

**Comment by glenvt18** [comment]
@mkaluza Those numbers look very nice. Great work! Thanks a lot for your contribution.

**Pull request merged**